python/services/dataservice/dmod/dataservice/initial_data_adder_impl.py

```diff
@@ -0,0 +1,480 @@
+        # TODO: double check that this is being added when we do data checks
+        forcing_req = [r for r in self._job_request.data_requirements if r.category == DataCategory.FORCING][0]
+        forcing_dataset_name = forcing_req.fulfilled_by
+        manager = [m for _, m in self._all_dataset_managers.items() if forcing_dataset_name in m.datasets][0]
+        forcing_dataset = manager.datasets[forcing_dataset_name]
+
+        # Figure out the correct provider type from the dataset format
+        # TODO: this may not be the right way to do this to instantiate the object directly (i.e., not through JSON)
+        if forcing_dataset.data_format == DataFormat.NETCDF_FORCING_CANONICAL:
+            forcing_cfg_params['provider'] = 'NetCDF'
+        elif forcing_dataset.data_format == DataFormat.AORC_CSV:
+            forcing_cfg_params['provider'] = 'CsvPerFeature'
+
+        # TODO: (#needs_issue) introduce logic to examine forcing dataset and intelligently assess what the file
+        #  name(s)/pattern(s) should be if they aren't explicitly provided
+
+        if self.partial_realization_config is not None and self.partial_realization_config.forcing_file_pattern:
+            forcing_cfg_params['file_pattern'] = self.partial_realization_config.forcing_file_pattern
+
+        # Finally, produce the right path
+        # TODO: these come from scheduler.py; may need to centralize somehow
+        forcing_cfg_params['path'] = '/dmod/datasets/'
+        if self.partial_realization_config is not None and self.partial_realization_config.is_env_workaround:
+            forcing_cfg_params['path'] += 'from_env'
+        else:
+            forcing_cfg_params['path'] += '{}/{}/'.format(DataCategory.FORCING.name.lower(), forcing_dataset_name)
+
+        if self.partial_realization_config is not None and self.partial_realization_config.forcing_file_name:
+            forcing_cfg_params['path'] += self.partial_realization_config.forcing_file_name
+
+        return Forcing(**forcing_cfg_params)
+
+    def build_realization_config_from_partial(self) -> NgenRealization:
+        """
+        Build a ngen realization config object from current service state and partial config within the job request.
+
+        Returns
+        -------
+        NgenRealization
+            The built realization config.
+        """
+        params = dict()
+
+        if self.partial_realization_config.global_formulations is not None:
+            params['global_config'] = Realization(formulations=self.partial_realization_config.global_formulations,
+                                                  forcing=self._build_forcing_config_for_realization())
+
+        params['time'] = Time(start_time=self._job_request.time_range.begin, end_time=self._job_request.time_range.end)
+
+        if self.partial_realization_config.routing_config is not None:
+            params['routing'] = self.partial_realization_config.routing_config
```

Comment:
I wasnt specifically thinking of the chicken and egg problem, although my question is related. Sorry for not communicating more effectively, let me try and re-ask. Do we need to modify the path the routing configuration file here, if it is provided, similar to what is done in `_build_forcing_config_for_realization`?

```python
        forcing_cfg_params['path'] = '/dmod/datasets/'
        if self.partial_realization_config is not None and self.partial_realization_config.is_env_workaround:
            forcing_cfg_params['path'] += 'from_env'
```

So for example, check that it is a relative path and make any corrections to that path if it is not relative to the realization config dataset path.

Reply:
Yeah, I was approaching the same fundamental problem from the other side of it.  

Yes, we probably do need to normalize path values for any created config dataset so that the paths will be valid inside the eventual worker container.  Which then leads to the type of problem I alluded to.

Reply:
Is that something that you'd like to address in this PR? Or just TODO and issue for later?

Reply:
We need to handled that separately.  I've opened #407 for that.  

We may also need a separate issue (perhaps on the _ngen-cal_/_ngen-config_ side of things) for dealing with validations issues for some of the config objects we are making use of in situations when paths aren't there.

Reply:
> We may also need a separate issue (perhaps on the _ngen-cal_/_ngen-config_ side of things) for dealing with validations issues for some of the config objects we are making use of in situations when paths aren't there.

Great point. This is something Ive been thinking a lot about. There are several field types, file and directory paths are maybe the best example, where it is desirable in some cases to validate and deserialize in a stricter mode. So, in strict mode, you would want to ensure that the paths exist for example. I think ive come up with a good solution to do this, I just need to propose it in the `ngen.config` package. 

In the short term, we could always add a utility function that walks a pydantic model's fields and if a field is a Path type, then verify that it exists. Something like:


```python
from pydantic import BaseModel
from pathlib import Path
from typing import Type, List

def verify_path_types_exists(m: BaseModel) -> List[Path]:
  horizon: List[BaseModel] = [m]
  errors = []

  while horizon:
    model = horizon.pop()
    for field_name in model.__fields__:
      field_value = getattr(model, field_name)
      if isinstance(field_value, Path):
        if not field_value.exists():
          errors.append(field_value)
      elif isinstance(field_value, BaseModel):
        horizon.append(field_value)

  return errors
```

Reply:
I'm curious about your solution.  I was under the impression there wasn't a way around this without changing the type for _DirectoryPath_ and _FilePath_ attributes, potentially in some kind of dynamic way (i.e., a "strict" mode).  

I don't follow how the short-term solution will prevent the validation problems (unless we are again talking about changing the attribute types).

Reply:
> I'm curious about your solution. I was under the impression there wasn't a way around this without changing the type for _DirectoryPath_ and _FilePath_ attributes, potentially in some kind of dynamic way (i.e., a "strict" mode).

I guess you could do it using some really complex try catch stuff and `.construct()`. I accomplished it by [doing something closer](https://github.com/NOAA-OWP/t-route/blob/cd51363ebcf03efcb98014bdf206b89a10fb55ec/src/troute-config/troute/config/types.py) to what you are suggesting.

> I don't follow how the short-term solution will prevent the validation problems (unless we are again talking about changing the attribute types).

I _think_ file path and directory path field types in `ngen.config` are not currently strictly for this reason. So, it will in all cases make sure to coerce the input to a `Path`, but I don't think it fails if the file or directory doesnt exist.